File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n = prices.size();
        vector<vector<int>> dp(n + 1, vector<int>(2));
        dp[n][0] = 0, dp[n][1] = 0;

        for (int idx = n - 1; idx >= 0; idx--) {
            for (int canBuy = 1; canBuy >= 0; canBuy--) {
                int buy = INT_MIN, sell = INT_MIN;
                if (canBuy) {
                    buy = -prices[idx] + dp[idx + 1][!canBuy];
                } else {
                    sell = prices[idx] - fee + dp[idx + 1][!canBuy];
                }
                int skip = dp[idx + 1][canBuy];
                dp[idx][canBuy] = max({buy, sell, skip});
            }
        }

        return dp[0][1];
    }
};

File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        n = len(prices)

        memo = {}

        def solve(index, isBuy, isSell):
            if index == n:
                return 0
            key = (index, isBuy, isSell)
            if key in memo:
                return memo[key]

            buy = sell = skip = 0

            if not isBuy:
                buy = (-1 * prices[index]) + solve(index + 1, True, False)
            elif not isSell:
                sell = (-1 * fee) + prices[index] + solve(index + 1, False, True)

            skip = solve(index + 1, isBuy, isSell)

            memo[key] = max(buy, sell, skip)
            return memo[key]

        return solve(0, False, False)
